### PR TITLE
[Data] Add partitioning classes to Data API reference

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

#23624 added support for path partitioning, but didn't add documentation to the API reference.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/package-ref.rst [hunk 1]
```diff
@@ -122,6 +122,21 @@ Custom Datasource API
 .. autoclass:: ray.data.ReadTask
     :members:
 
+Partitioning API
```
Comment (pdames) [inline doc/source/data/package-ref.rst:125]:
```suggestion
Datasource Partitioning API
```

To distinguish this from a Dataset Partitioning API (which may be introduced later)?

File: doc/source/data/package-ref.rst [hunk 2]
```diff
@@ -122,6 +122,21 @@ Custom Datasource API
 .. autoclass:: ray.data.ReadTask
     :members:
 
+Datasource Partitioning API
+---------------------
+
+.. autoclass:: ray.data.datasource.PartitionStyle
+    :members:
+
```
Comment (pdames) [inline doc/source/data/package-ref.rst:130]:
```suggestion

.. autoclass:: ray.data.datasource.PathPartitionScheme
    :members:
    
```
(unless there's a reason for `PathPartitionScheme` to be intentionally omitted)

Comment (bveeramani) [reply doc/source/data/package-ref.rst:130]: Fixed!

**Review by ericl — COMMENTED** [review]
So these classes actually aren't in the public api of Dataset. I think we need to decide whether they should be public or not. If they're not public/dev api, then we shouldn't document them, and furthermore should move them into a `impl` directory.

My vote is to mark all of them as DeveloperAPI, but would leave it to @clarkzinzow and @jianoaix to decide.

**Comment by jianoaix** [comment]
> So these classes actually aren't in the public api of Dataset. I think we need to decide whether they should be public or not. If they're not public/dev api, then we shouldn't document them, and furthermore should move them into a `impl` directory.
> 
> My vote is to mark all of them as DeveloperAPI, but would leave it to @clarkzinzow and @jianoaix to decide.

Yeah, I had similar question. The issue is that those classes are used as part of the public APIs, e.g. read_parquet() will take a parquet metadata provider as parameter. It seems transitively make those classes public.

**Comment by clarkzinzow** [comment]
Yeah @ericl they are part of the public `read_*` APIs, so I'd vote that they be considered public APIs.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
LINT failure

**Comment by jianoaix** [comment]
@bveeramani can you fix the failure and make this PR merge-able? We'd like to get it picked to 1.13, given this is just doc fix and public API docstring is needed.

**Comment by bveeramani** [comment]
> @bveeramani can you fix the failure and make this PR merge-able? We'd like to get it picked to 1.13, given this is just doc fix and public API docstring is needed.

Fixed the lint failure. I took a look at the linkcheck failure and to the best of my knowledge the failure is unrelated to this PR.

**Pull request merged**